### PR TITLE
Fast finish AppVeyor (Pt. 2)

### PR DESCRIPTION
**Pull request by jakirkham**
Continues work that was done in PR ( https://github.com/conda-forge/staged-recipes/pull/332 ).

Basically it still feels to me that builds on `master` on AppVeyor are taking too long (~5 mins). So trying to speed this up more. Also, the recent issues downloading packages from `defaults` ( https://github.com/Anaconda-Platform/support/issues/49 ) are a bit concerning. Removing downloads from merged PRs at staged-recipes in AppVeyor builds should help mitigate that issue.

Also we have taken this opportunity to break up the `appveyor.yml` file into components. For instance, there is a `build_windows.bat`, which should hopefully be useful for offline builds (read debugging). There is `appveyor_setup.bat` as well, which wraps up our Windows setup nicely. It may be useful to have this batch file for configure other Windows VMs in the future.

cc @pelson @ocefpaf @msarahan @patricksnape @jjhelmus @gillins

### CODE REVIEW

**Comment by jakirkham** [comment]
Yeah, ignore the linter. Obviously there are no recipes. Though I might add some test PRs based off of this.

**Comment by jakirkham** [comment]
To get an idea of how much faster this builds, let's make some comparisons. Here is a recent [AppVeyor build]( https://ci.appveyor.com/project/conda-forge/staged-recipes/build/1.0.4982 ) from a merged recipe. Note each build takes ~1.5mins (little longer though it seems) and the total run for the full matrix is ~6mins. Considering that we can run two builds at a time and staged-recipes normally gets priority, we are looking at ~0.5mins that is potentially spent starting up (though probably less due to our underestimation).

With the changes we proposed here, I changed the remote to my fork and started this [AppVeyor build]( https://ci.appveyor.com/project/jakirkham/staged-recipes/build/1.0.36 ). As you can, see each build took ~5secs (in some cases less). There doesn't seem to be much room to make that any faster (as we can't skip them completely). The total run time was ~1.25mins. Though I don't have the nicer AppVeyor infrastructure that conda-forge has. So I can only run 1 build at a time and have a longer startup time. Based on this data that would be ~0.75mins spent starting up. When switching to AppVeyor, I would expect this startup time drops to ~0.5mins as well. Also, we would be able to run 2 builds at a time. This would mean the typical runtime would be ~0.75mins. This is a ~8 times improvement in speed. Though as noted the estimated build times for the current strategy are a little low and for this improvement a little high.

In short, this should yield very close to an order of magnitude speedup!

**Comment by jakirkham** [comment]
To verify things still build correctly on AppVeyor, I added the [`pycrypto` recipe]( https://github.com/jakirkham/staged-recipes/blob/2b268d4828556cc40fccef37dc8544f9763f54c0/recipes/pycrypto/meta.yaml ) back on top of these changes. The results can be seen in this [build]( https://ci.appveyor.com/project/jakirkham/staged-recipes/build/1.0.37 ). In short, it was able to build the full matrix without issues.

**Comment by jakirkham** [comment]
AFAICT this seems to be working. Though I would very much appreciate if some Windows/AppVeyor experts could check and see if what I'm doing is sound. Also it would be great if @pelson could verify that the call of `obvci_appveyor_python_build_env.cmd` is still working as intended given this rearrangement. Happy to work with people in trying some more tests if needed.

**Comment by jakirkham** [comment]
Should add this is only labeled WIP as I want to check with some people about whether this is working ok or it needs changes. If people are generally happy with it, I'm alright going ahead.

File: scripts/build_windows.bat [hunk 1]
```diff
@@ -0,0 +1,10 @@
+:: Workaround for https://github.com/conda/conda-build/issues/636
+set "PYTHONIOENCODING=UTF-8"
+
+:: SDK v7.0 MSVC Express 2008's SetEnv.cmd script will fail if the
+:: /E:ON and /V:ON options are not enabled in the batch script interpreter
+:: See: http://stackoverflow.com/a/13751649/163740
+set "CMD_IN_ENV=call cmd /E:ON /V:ON /C obvci_appveyor_python_build_env.cmd"
```
Comment (patricksnape) [inline scripts/build_windows.bat]: I'm still not 100% sure we need this (if it does what I think it does).

Comment (jakirkham) [inline scripts/build_windows.bat]: Here is the [script]( https://github.com/pelson/Obvious-CI/blob/v0.6.1/scripts/obvci_appveyor_python_build_env.cmd ). Maybe that will make it easier to determine.

Comment (patricksnape) [inline scripts/build_windows.bat]: Yes so this is activating the correct Windows SDK - which is only really required for Python 3.4 x64 and I think this is now correctly handled by [conda-build](https://github.com/conda/conda-build/blob/master/conda_build/windows.py#L142).

Comment (jakirkham) [inline scripts/build_windows.bat]: @pelson, how do you feel about dropping this?

Comment (jjhelmus) [inline scripts/build_windows.bat]: I agree with @patricksnape, I do not think this script is necessary any longer as `conda build` now performs the necessary activation of the SDKs and setting of the environment parameter.

**Comment by jjhelmus** [comment]
I like the idea of this change.  If `CMD_IN_ENV` is removed is `obvious-ci` even needed anymore?

**Comment by jakirkham** [comment]
> I like the idea of this change.

Awesome. Glad you like it.

> If `CMD_IN_ENV` is removed is `obvious-ci` even needed anymore?

My understanding is the answer is [no]( https://github.com/conda-forge/conda-smithy/issues/197#issuecomment-232021314 ). The plan was to add it to the `toolchain` ( https://github.com/conda-forge/toolchain-feedstock/issues/1 ) if it was needed. Though it sounds like it is not needed.

**Comment by jakirkham** [comment]
Even though it does sound like we are in agreement about removing the obvious-ci script, I'd like to postpone that change for a subsequent PR if that's alright. That being said, I'm in favor of that change happening if it truly is not needed.

**Comment by jakirkham** [comment]
Rebased after merging PR ( https://github.com/conda-forge/staged-recipes/pull/1143 ) to simplify this a little bit.

**Comment by jakirkham** [comment]
Removed the WIP. Would be nice to hear from @pelson about this change. Though feel free to use your own judgement as to whether it should go in yet or not.

File: scripts/build_windows.bat (hunk 1 above)
Comment (mwcraig) [inline scripts/build_windows.bat]: FWIW, I tried pulling it out on a different repo last week and the builds failed until I put it back.

Comment (jakirkham) [inline scripts/build_windows.bat]: That's a very interesting data point. Thanks for sharing, @mwcraig. I'd be curious if you could try this way of building on Windows/AppVeyor for that same use case to see if this works or needs more tweaks.

**Comment by mwcraig** [comment]
Sure, but it may be a little while because I'm on the road. Should have also mentioned that it was a few releases ago, maybe 1.21.5?

**Comment by jakirkham** [comment]
Of course. When you have time. Going to be out for a bit myself.

File: scripts/build_windows.bat (hunk 1 above)
Comment (patricksnape) [inline scripts/build_windows.bat]: @mwcraig Do you have an example of the issues caused by pulling it out? For which Python version and under what scenario? A link to the Appveyor perhaps?

**Comment by mwcraig** [comment]
@patricksnape -- I was trying to get a build working; the last fail it had is [here](https://ci.appveyor.com/project/Astropy/conda-channel-astropy/build/1.0.150#L825) (`cl.exe not found`). After that I made [this commit](https://github.com/astropy/conda-channel-astropy/commit/2611ae03acfe882682cf59af2ff87d289f689656), which restored `CMD_IN_ENV`. I had deleted it a few commits earlier as part of a string of attempts to get a build working.

After the commit restoring `CMD_IN_ENV`, the [build succeeded](https://ci.appveyor.com/project/Astropy/conda-channel-astropy/build/1.0.152).

All of this was with a slightly modified `conda-build=1.21.5` (the modification has since been merged).

Haven't tried again with a more recent version...

**Comment by patricksnape** [comment]
@msarahan Just out of interest, what forced you to do [this](https://github.com/conda/conda-build/commit/7c20b8578b88c3271b747b6b9edbfb8b35c6f332)? It looks like we aren't using the Python tools at all now?

**Comment by jakirkham** [comment]
To steer the ship 🚢 a little, are these changes as is acceptable/mergeable? If so, let's merge them. If not, it would be great to know what still needs to change. We can continue the Obvious-CI configuration script discussion in a subsequent PR.

**Comment by jakirkham** [comment]
Rebased and handled conflicts created by PR ( https://github.com/conda-forge/staged-recipes/pull/1373 ).

Also re-ran test builds [1]( https://ci.appveyor.com/project/jakirkham/staged-recipes/build/1.0.379/job/jwbr81t0d8002i1f ), [2]( https://ci.appveyor.com/project/jakirkham/staged-recipes/build/1.0.380 ), and [3]( https://ci.appveyor.com/project/jakirkham/staged-recipes/build/1.0.381 ). Builds 1 and 3 just verify that things behave normally both without and with a recipe (`pycrypto`) that requires some compilation. Build 2 verifies the fast finish will work and gives an idea of how fast it is.

**Comment by jakirkham** [comment]
Could you please take a look at this @pelson and merge if you are happy? I think this will be very helpful with our AppVeyor performance at staged-recipes. Plus it is mostly copying and pasting code around so there isn't actually that much going on in it. Already have done a fair bit of testing myself and am confident it works.

**Comment by jakirkham** [comment]
I think the trivial change made in PR ( https://github.com/conda-forge/staged-recipes/pull/1419 ) obviates the need for this fast finish. That PR simply does not run AppVeyor builds when merged. There still may be things worth salvaging from this PR. So will leave it open for the time being. However, am not planning to pursue it further at this time.
